frb_codegen/src/parser/mod.rs

```diff
@@ -16,9 +18,73 @@ use crate::method_utils::FunctionName;
 use crate::parser::ty::TypeParser;
 use crate::source_graph::Crate;
 
+use self::ty::convert_ident_str;
+
 const STREAM_SINK_IDENT: &str = "StreamSink";
 const RESULT_IDENT: &str = "Result";
 
+pub(crate) fn get_toposort(src: HashMap<String, Type>) -> HashMap<String, Type> {
+    // Some types that cannot be Handled.
+    // Filter something like `BareFn( TypeBareFn...`
+    // Filter something like `Ptr( TypePtr { star_token: Star,`
+    let mut ret: HashMap<String, Type> = src
+        .iter()
+        .filter_map(|(k, v)| match convert_ident_str(v) {
+            Some(_) => None,
+            None => Some((k.to_owned(), v.to_owned())),
+        })
+        .collect();
+
+    let string_src = src
+        .iter()
+        // Filter some types that cannot be Handled.
+        .filter_map(|(k, v)| convert_ident_str(v).map(|v| (k.to_owned(), v)));
+
+    let mut ts = TopologicalSort::<String>::new();
+
+    string_src.for_each(|(k, v)| {
+        // k and v switch orders here
+        ts.add_dependency(v, k);
+    });
+
+    // remove base type, like i32 in `type Id = i32`.
+    // You might worry about the type, which cannot be handled and isn't in ts.
+    // Case:
+    // ```
+    // type Id = unsafe{};
+    // type UserId = Id;
+    // ```
+    // 1. pop_base: ret = [{Id, unsafe}]
+    // 2. init_condition: pop Id, and then handle UserId.
+    //    src.get("UserId") => Id,
+    ts.pop_all();
+    // build init_condition
+    ts.pop_all().into_iter().for_each(|k| {
+        let v_src = src.get(&k).unwrap().to_owned();
+        let v_str = convert_ident_str(&v_src).unwrap();
+
+        ret.insert(
+            k,
+            if ret.contains_key(&v_str) {
+                ret.get(&v_str).unwrap().clone()
+            } else {
+                v_src
+            },
+        );
+    });
+
+    while let Some(k) = ts.pop() {
+        let v_src = src.get(&k).unwrap().to_owned();
+        let v_str = convert_ident_str(&v_src).unwrap();
+        let v_ret = ret
+            .get(&v_str)
+            .unwrap_or_else(|| panic!("{:?},\n{:?},\n{:?},\n{}", src, ts, ret, k))
+            .to_owned();
+        ret.insert(k, v_ret);
+    }
+    ret
```

Comment:
uber-nit: Is it possible we merge the two sub-sections into one

Reply:
* in init_condition
```
if ret.contains_key(&v_str) {
    // only happen if v_src cannot handle  <<<<<
    ret.get(&v_str).unwrap().clone()
}
```
So, it is dealing with primitive types like i32 most time

* not in init_condition
But after that, you don't have to worry about the boundary case, you can just merge the dependencies

So, I didn't know how to merge them -_-||